course/src/isometry.cc

```diff
@@ -0,0 +1,124 @@
+#include "isometry.h"
+#include <cmath>
+#include "matrix3.h"
+#include "vector3.h"
+
+namespace ekumen {
+namespace math {
+namespace {}  // namespace
+
+Isometry::Isometry(const Vector3& translation, const Matrix3& rotation)
+    : translation_(translation), rotation_(rotation) {}
+
+Isometry::Isometry(const Matrix3& rotation)
+    : translation_(Vector3::kZero), rotation_(rotation) {}
+
+Isometry::Isometry(const Isometry& obj)
+    : Isometry(obj.translation_, obj.rotation_) {}
+
+Isometry::Isometry(Isometry&& obj)
+    : translation_(std::move(obj.translation_)),
+      rotation_(std::move(obj.rotation_)) {}
+
+Isometry& Isometry::operator=(const Isometry& obj) {
+  translation_ = obj.translation_;
+  rotation_ = obj.rotation_;
+  return *this;
+}
+Isometry& Isometry::operator=(Isometry&& obj) {
+  // The rvalue reference shouldn't be the same as this.
+  if (this == &obj) {
+    throw std::invalid_argument(
+        "rvalue cannot be identity of lvalue in move assignment.");
+  }
+
+  translation_ = std::move(obj.translation_);
+  rotation_ = std::move(obj.rotation_);
+  return *this;
+}
+
+Isometry Isometry::FromTranslation(const Vector3& translation) {
+  return Isometry(translation);
+}
+
+Isometry Isometry::RotateAround(const Vector3& axis, const double& angle) {
+  Matrix3 res(Matrix3::kIdentity);
+  double cosAngle = std::cos(angle);
+  double sinAngle = std::sin(angle);
+  if (axis == Vector3::kUnitX) {
+    res[1][1] = cosAngle;
+    res[1][2] = -1 * sinAngle;
+    res[2][1] = sinAngle;
+    res[2][2] = cosAngle;
+  } else if (axis == Vector3::kUnitY) {
+    res[0][0] = cosAngle;
+    res[0][2] = sinAngle;
+    res[2][0] = -1 * sinAngle;
+    res[2][2] = cosAngle;
+  } else if (axis == Vector3::kUnitZ) {
+    res[0][0] = cosAngle;
+    res[0][1] = -1 * sinAngle;
+    res[1][0] = sinAngle;
+    res[1][1] = cosAngle;
+  } else {
+    throw std::invalid_argument(
+        "Axis to rotate around should be a unitary versor.");
+  }
+  return Isometry(res);
+}
+
+Isometry Isometry::FromEulerAngles(const double& psi, const double& theta,
+                                   const double& phi) {
+  Isometry psi_rotation = Isometry::RotateAround(Vector3::kUnitX, psi);
+  Isometry theta_rotation = Isometry::RotateAround(Vector3::kUnitY, theta);
+  Isometry phi_rotation = Isometry::RotateAround(Vector3::kUnitZ, phi);
+  return psi_rotation * theta_rotation * phi_rotation;
+}
+
+Matrix3 Isometry::rotation() const { return rotation_; }
+Vector3 Isometry::translation() const { return translation_; }
+
+Isometry Isometry::operator*(const Isometry& obj) const {
+  double res_translation_x =
+      rotation_.row(0).dot(obj.translation_) + translation_.x();
+  double res_translation_y =
+      rotation_.row(1).dot(obj.translation_) + translation_.y();
+  double res_translation_z =
+      rotation_.row(2).dot(obj.translation_) + translation_.z();
+
+  Vector3 res_translation(res_translation_x, res_translation_y,
```

Comment:
While this seems mathematically correct, please make sure you implemente the `Matrix3 * Vector3` product in the `Matrix3` class, and then the translation of `I_1 * I_2` can be written as `R_1 * t_2 + t_1`

Reply:
Right! Done.